### PR TITLE
feat(web): refactor gRPC to HTTP API

### PR DESCRIPTION
**Pull request by robot9706**
Refactored the following gRPC (`crux` and `crux-ui`) APIs to REST:

- Audit
- Dashboard
- Product
- Registry
- Token
- Version

Co-authored-by: Mate Vago < @m8vago >
Co-authored-by: Levente Orban < @polaroi8d >

### CODE REVIEW

File: web/crux/src/app/product/product.dto.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+export class BasicProductDto extends DateAuditProperties {
+  @IsUUID()
+  id: string
+
+  @IsString()
+  name: string
+
+  @IsString()
+  @IsOptional()
+  description?: string | undefined
```
Comment (polaroi8d) [inline web/crux/src/app/product/product.dto.ts:19]: Why we need undefined?

**Review by m8vago — CHANGES_REQUESTED** [review]

File: web/crux-ui/src/models/audit.ts [hunk 2]
```diff
@@ -1,3 +1,10 @@
+export type Audit = {
+  updatedAt?: string | null
```
Comment (m8vago) [inline web/crux-ui/src/models/audit.ts:2]: this will never be null

File: web/crux-ui/src/routes.ts [hunk 3]
```diff
@@ -146,12 +146,12 @@ export const verificationUrl = (email: string) => `${ROUTE_VERIFICATION}?email=$
...
 export const productUrl = (id: string, params?: VersionUrlParams) => appendUrlParams(`${ROUTE_PRODUCTS}/${id}`, params)
-export const productApiUrl = (id: string) => `/api${productUrl(id)}`
-export const productVersionsApiUrl = (productId: string) => `${productApiUrl(productId)}/versions`
+export const productApiUrl = (id: string) => `${API_PRODUCTS}/${id}`
+export const productVersionsApiUrl = (productId: string) => `${productApiUrl(productId)}/${productId}/versions`
 
 // registry
 export const registryUrl = (id: string) => `${ROUTE_REGISTRIES}/${id}`
-export const registryApiUrl = (id: string) => `/api${registryUrl(id)}`
+export const registryApiUrl = (id: string) => `/api/new${registryUrl(id)}`
```
Comment (m8vago) [inline web/crux-ui/src/routes.ts:154]: `${API_REGISTRIES}/${id}`

File: web/crux/src/app/registry/registry.dto.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
+import { ApiProperty, getSchemaPath } from '@nestjs/swagger'
+import { Type } from 'class-transformer'
+import { IsBoolean, IsDate, IsIn, IsOptional, IsString, IsUrl, IsUUID } from 'class-validator'
+
+export const REGISTRY_TYPE_VALUES = ['v2', 'hub', 'gitlab', 'github', 'google', 'unchecked'] as const
+export type RegistryType = (typeof REGISTRY_TYPE_VALUES)[number]
+
+const RegistryDetailOneOf = () =>
```
Comment (m8vago) [inline web/crux/src/app/registry/registry.dto.ts:8]: Details*

File: web/crux/src/app/registry/registry.http.controller.ts [hunk 5]
```diff
@@ -1,28 +1,109 @@
...
+  // TODO(@robot9706): Refactor the auditlog after removing gRPC
+  @AuditLogLevel('disabled')
+  @ApiOkResponse({ type: RegistryDetails })
+  async getRegistry(@RegistryId() id: string): Promise<RegistryDetails> {
+    return await this.service.getRegistryDetails(id)
+  }
+
+  @Post()
+  @UseGuards(RegistryAccessValidationGuard)
+  @AuditLogLevel('disabled')
```
Comment (m8vago) [inline web/crux/src/app/registry/registry.http.controller.ts:61]: unnecessary

File: web/crux/src/app/token/token.http.controller.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+  @Get(':tokenId')
+  @ApiOkResponse({ type: Token })
+  async getToken(@TokenId() id: string, @IdentityFromRequest() identity: Identity): Promise<Token> {
+    return this.service.getToken(id, identity)
+  }
+
+  @Post()
+  @ApiBody({ type: GenerateToken })
+  @ApiCreatedResponse({
```
Comment (m8vago) [inline web/crux/src/app/token/token.http.controller.ts:58]: create a custom decorator for this like the @TokenId up there

File: web/crux/src/app/token/token.mapper.ts [hunk 7]
```diff
@@ -1,21 +1,18 @@
...
-import { Token } from '@prisma/client'
-import { toTimestamp } from 'src/domain/utils'
-import { TokenResponse } from 'src/grpc/protobuf/proto/crux'
+import { Token as PrismaToken } from '@prisma/client'
+import { Token } from './token.dto'
 
 @Injectable()
 export default class TokenMapper {
-  toGrpc(token: Token): TokenResponse {
+  listItemToDto(token: PrismaToken): Token {
```
Comment (m8vago) [inline web/crux/src/app/token/token.mapper.ts:7]: toDto

File: web/crux/src/app/token/token.mapper.ts [hunk 8]
```diff
@@ -1,21 +1,18 @@
...
     return {
       ...token,
-      expiresAt: toTimestamp(token.expiresAt),
-      createdAt: toTimestamp(token.createdAt),
     }
   }
 
-  generateResponseToGrpc(prismaToken: Token, jwt: string) {
+  generateResponseToDto(prismaToken: PrismaToken, jwt: string) {
```
Comment (m8vago) [inline web/crux/src/app/token/token.mapper.ts:13]: missing return type

File: web/crux/src/app/version/version.mapper.ts [hunk 9]
```diff
@@ -1,40 +1,68 @@
...
       deletable: versionIsDeletable(version),
       increasable: versionIsIncreasable(version),
-      images: version.images.map(it => this.imageMapper.detailsToProto(it)),
-      deployments: version.deployments.map(it => this.deployMapper.deploymentByVersionToProto(it)),
+      images: version.images.map(it => ({
+        id: it.id,
+        name: it.name,
+        tag: it.tag,
+        order: it.order,
+        config: it.config as any as ContainerConfigData,
```
Comment (m8vago) [inline web/crux/src/app/version/version.mapper.ts:47]: replace this with ContainerConfigDto

**Review by m8vago — APPROVED** [review]

**Review by polaroi8d — APPROVED** [review]

**Pull request merged**